bugfix: expected HTTP codes / error classes / error messages shouldn't impact Apdex (#2619)
newrelic/newrelic-ruby-agent: CHANGELOG.md

# New Relic Ruby Agent Release Notes

## dev

> Version <dev> fixes a bug that would cause an error and a correspondong expected HTTP status code to negatively impact a transaction's Apdex.

Reviewer: ```suggestion
Version <dev> fixes a bug that would cause an error and a corresponding expected HTTP status code to negatively impact a transaction's Apdex.
```

> Version <dev> fixes a bug that would cause an error and a corresponding expected HTTP status code to negatively impact a transaction's Apdex.

Reviewer 2: the way this sentence is worded to me sounds like its only about errors related to status codes, rather than all expected errors. My understanding is this is more about expected errors in general than the http status codes config (correct me if i'm wrong of course!). maybe this could be reworded for clarity?
something like this maybe?
```suggestion
Version <dev> fixes a bug that would cause an expected error to negatively impact a transaction's Apdex.
```

Author: Great call, @tannalynn. The fix includes routing the "should this error impact apdex?" check through a complete "is the error expected?" check, which will indeed check status, class, and message. But I got tunnel vision when it came to writing the text description.

> - **Bugfix: Expected HTTP status code / error class / error message based errors won't impact Apdex

Reviewer: ```suggestion
- **Bugfix: Expected errors related to HTTP status code, class, and message won't impact Apdex
```

>   The agent is supposed to prevent observed application errors from negatively impacting Apdex if the errors are to be either ignored or expected. There are 2 ways for the agent to expect an error: via the `notice_error` API receiving an `expected: true` argument or via a matches made against the configuration file based user defined lists for expected HTTP status codes (`:'error_collector.expected_status_codes'`), expected error classes (`:'error_collector.expected_classes'`), or expected error messages (`:'error_collector.expected_messages'`). Previously only errors expected via the `notice_error` API were correctly prevented from impacting Apdex and the configuration parameter based expected errors were incorrectly impacting Apdex. This behavior has been fixed and now both types of expected errors will correctly not impact Apdex. Thanks very much to [@florianpilz](https://github.com/florianpilz) for bringing this issue to our attention. [PR#2619](https://github.com/newrelic/newrelic-ruby-agent/pull/2619)

Reviewer: ```suggestion
  The agent is supposed to prevent observed application errors from negatively impacting Apdex if the errors are either ignored or expected. There are two ways for the agent to expect an error: via the `notice_error` API receiving an `expected: true` argument or via matches made against user-configured lists for expected HTTP status codes (`:'error_collector.expected_status_codes'`), expected error classes (`:'error_collector.expected_classes'`), or expected error messages (`:'error_collector.expected_messages'`). Previously, only errors expected via the `notice_error` API were correctly prevented from impacting Apdex. Expected errors set by configuration incorrectly impacted Apdex. This behavior has been fixed and now both types of expected errors will correctly not impact Apdex. Thanks very much to [@florianpilz](https://github.com/florianpilz) for bringing this issue to our attention. [PR#2619](https://github.com/newrelic/newrelic-ruby-agent/pull/2619)
```
